packages/@aws-cdk/integ-runner/lib/runner/integration-tests.ts

```diff
@@ -248,15 +296,33 @@ export class IntegrationTests {
   }
 
   private async discover(options: IntegrationTestsDiscoveryOptions): Promise<IntegTest[]> {
-    const patterns = options.testRegex ?? ['^integ\\..*\\.js$'];
+    const languagePresets = options.language ?? Array.from(IntegrationTests.defaultDiscoveryRegexes.keys());
+    const patterns = options.testRegex?.map((pattern) => new RegExp(pattern))
+        ?? Array.from(IntegrationTests.defaultDiscoveryRegexes.entries()).filter(
+          ([language]) => languagePresets.includes(language),
+        ).map(([_, regex]) => regex);
 
     const files = await this.readTree();
-    const integs = files.filter(fileName => patterns.some((p) => {
-      const regex = new RegExp(p);
+    const integs = files.filter(fileName => patterns.some((regex) => {
       return regex.test(fileName) || regex.test(path.basename(fileName));
     }));
 
-    return this.request(integs, options);
+    const discoveredTestNames = new Set<string>();
+    const integsWithoutDuplicates = new Array<string>();
+
+    // Remove tests with duplicate names.
+    // To make sure the precendence of files is deterministic, iterate the files in lexicographic order.
+    // Additionally, to give precedence to compiled .js files over their .ts source,
+    // use ascending lexicographic ordering, so the .ts files are picked up first.
+    for (const integFileName of integs.sort()) {
+      const testName = path.parse(integFileName).name;
+      if (!discoveredTestNames.has(testName)) {
+        integsWithoutDuplicates.push(integFileName);
+      }
+      discoveredTestNames.add(testName);
+    }
```

Comment:
I wonder if still need a generic de-duplication or just a special case `if languages includes 'typescript' AND 'javascript'`

Either way could we reduce duplication if we added the files to be ignored to `options.exclude` and let `filterTests()` do its work?